test/py/libvfio_user.py

```diff
@@ -685,25 +685,42 @@ def connect_sock():
     return sock
 
 
-def connect_client(ctx):
-    sock = connect_sock()
-
-    json = b'{ "capabilities": { "max_msg_fds": 8 } }'
-    # struct vfio_user_version
-    payload = struct.pack("HH%dsc" % len(json), LIBVFIO_USER_MAJOR,
-                          LIBVFIO_USER_MINOR, json, b'\0')
-    hdr = vfio_user_header(VFIO_USER_VERSION, size=len(payload))
-    sock.send(hdr + payload)
-    vfu_attach_ctx(ctx, expect=0)
-    payload = get_reply(sock, expect=0)
-    return sock
-
-
-def disconnect_client(ctx, sock):
-    sock.close()
-
-    # notice client closed connection
-    vfu_run_ctx(ctx, errno.ENOTCONN)
+class Client:
+    """Models a VFIO-user client connected to the server under test."""
+
+    def __init__(self, sock=None):
+        self.sock = sock
+        self.client_cmd_socket = None
+
+    def connect(self, ctx):
+        self.sock = connect_sock()
+
+        json = b'{ "capabilities": { "max_msg_fds": 8 } }'
+        # struct vfio_user_version
+        payload = struct.pack("HH%dsc" % len(json), LIBVFIO_USER_MAJOR,
+                              LIBVFIO_USER_MINOR, json, b'\0')
+        hdr = vfio_user_header(VFIO_USER_VERSION, size=len(payload))
+        self.sock.send(hdr + payload)
+        vfu_attach_ctx(ctx, expect=0)
+        fds, payload = get_reply_fds(self.sock, expect=0)
+        self.client_cmd_socket = socket.socket(fileno=fds[0]) if fds else None
```

Comment:
This will never be assigned until the `twin_sock` changes, right?

Reply:
Ah, good catch, these lines ended up in the wrong patch. I'll move them out of this one.